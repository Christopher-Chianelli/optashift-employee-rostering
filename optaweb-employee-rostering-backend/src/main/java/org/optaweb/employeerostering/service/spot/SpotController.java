/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.optaweb.employeerostering.service.spot;

import java.util.List;

import javax.enterprise.context.ApplicationScoped;
import javax.inject.Inject;
import javax.validation.Valid;
import javax.validation.constraints.Min;
import javax.ws.rs.DELETE;
import javax.ws.rs.GET;
import javax.ws.rs.POST;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;

import org.optaweb.employeerostering.domain.spot.Spot;
import org.optaweb.employeerostering.domain.spot.view.SpotView;

@Path("/rest/tenant/{tenantId}/spot")
@ApplicationScoped
// @Api(tags = "Spot")
public class SpotController {

    private final SpotService spotService;

    @Inject
    public SpotController(SpotService spotService) {
        this.spotService = spotService;
        // Assert.notNull(spotService, "spotService must not be null.");
    }

    // @ApiOperation("Get a list of all spots")
    @GET
    @Path("/")
    public List<Spot> getSpotList(@PathParam("tenantId") @Min(0) Integer tenantId) {
        return spotService.getSpotList(tenantId);
    }

    // @ApiOperation("Get a spot by id")
    @GET
    @Path("/{id}")
    public Spot getSpot(@PathParam("tenantId") @Min(0) Integer tenantId,
            @PathParam("id") @Min(0) Long id) {
        return spotService.getSpot(tenantId, id);
    }

    // @ApiOperation("Delete a spot")
    @DELETE
    @Path("/{id}")
    public Boolean deleteSpot(@PathParam("tenantId") @Min(0) Integer tenantId,
            @PathParam("id") @Min(0) Long id) {
        return spotService.deleteSpot(tenantId, id);
    }

    // @ApiOperation("Add a new spot")
    @POST
    @Path("/add")
    public Spot createSpot(@PathParam("tenantId") @Min(0) Integer tenantId,
            @Valid SpotView spotView) {
        return spotService.createSpot(tenantId, spotView);
    }

    // @ApiOperation("Update a spot")
    @POST
    @Path("/update")
    public Spot updateSpot(@PathParam("tenantId") @Min(0) Integer tenantId,
            @Valid SpotView spotView) {
        return spotService.updateSpot(tenantId, spotView);
    }
}
