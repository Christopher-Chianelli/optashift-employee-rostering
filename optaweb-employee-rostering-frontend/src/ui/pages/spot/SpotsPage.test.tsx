/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import { Spot } from 'domain/Spot';
import { getRouterProps } from 'util/BookmarkableTestUtils';
import { mockStore } from 'store/mockStore';
import { Skill } from 'domain/Skill';
import { Map } from 'immutable';
import DomainObjectView from 'domain/DomainObjectView';
import { spotSelectors } from 'store/spot';
import { mockRedux } from 'setupTests';
import { SpotsPage, SpotRow, EditableSpotRow } from './SpotsPage';

const noSpotsStore = mockStore({
  spotList: {
    isLoading: false,
    spotMapById: Map(),
  },
}).store;

const twoSpotsStore = mockStore({
  skillList: {
    isLoading: false,
    skillMapById: Map<number, Skill>()
      .set(0, {
        id: 0,
        version: 0,
        tenantId: 0,
        name: 'Skill 1',
      }),
  },
  spotList: {
    isLoading: false,
    spotMapById: Map<number, DomainObjectView<Spot>>()
      .set(1, {
        id: 1,
        version: 0,
        tenantId: 0,
        name: 'Spot 1',
        requiredSkillSet: [],
      })
      .set(2, {
        id: 2,
        version: 0,
        tenantId: 0,
        name: 'Spot 2',
        requiredSkillSet: [0],
      }),
  },
}).store;

describe('Spots page', () => {
  it('should render correctly with no spots', () => {
    mockRedux(noSpotsStore);
    const spotsPage = shallow(<SpotsPage {...getRouterProps('/0/spot', {})} />);
    expect(toJson(spotsPage)).toMatchSnapshot();
  });

  it('should render correctly with a few spots', () => {
    mockRedux(twoSpotsStore);
    const skillsPage = shallow(
      <SpotsPage {...getRouterProps('/0/spot', {})} />,
    );
    expect(toJson(skillsPage)).toMatchSnapshot();
  });

  it('should render the viewer correctly', () => {
    const spot = spotSelectors.getSpotById(twoSpotsStore.getState(), 2);
    mockRedux(twoSpotsStore);
    getRouterProps('/0/spot', {});
    const viewer = shallow(<SpotRow {...spot} />);
    expect(toJson(viewer)).toMatchSnapshot();
  });

  it('should render the editor correctly', () => {
    const spot: Spot = {
      tenantId: 0,
      id: 1,
      name: 'Spot',
      requiredSkillSet: [],
    };
    mockRedux(twoSpotsStore);
    const editor = shallow(<EditableSpotRow spot={spot} isNew={false} onClose={jest.fn()} />);
    expect(toJson(editor)).toMatchSnapshot();
  });
});
