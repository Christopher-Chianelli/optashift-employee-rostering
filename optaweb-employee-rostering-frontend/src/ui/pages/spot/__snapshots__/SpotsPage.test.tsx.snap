// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spots page should render correctly with a few spots 1`] = `
<TheTable
  columns={
    Array [
      Object {
        "name": "Trans(i18nKey=name)",
        "sorter": [Function],
      },
      Object {
        "name": "Trans(i18nKey=requiredSkillSet)",
      },
    ]
  }
  filterText=""
  filteredRows={
    Immutable.List [
      Object {
        "id": 1,
        "name": "Spot 1",
        "requiredSkillSet": Array [],
        "tenantId": 0,
        "version": 0,
      },
      Object {
        "id": 2,
        "name": "Spot 2",
        "requiredSkillSet": Array [
          Object {
            "id": 0,
            "name": "Skill 1",
            "tenantId": 0,
            "version": 0,
          },
        ],
        "tenantId": 0,
        "version": 0,
      },
    ]
  }
  history={
    Object {
      "action": "PUSH",
      "block": [MockFunction],
      "createHref": [MockFunction],
      "go": [MockFunction],
      "goBack": [MockFunction],
      "goForward": [MockFunction],
      "length": 1,
      "listen": [MockFunction],
      "location": Object {
        "hash": "",
        "pathname": "/0/spot",
        "search": "",
        "state": undefined,
      },
      "push": [MockFunction],
      "replace": [MockFunction],
    }
  }
  isReversed={true}
  itemsPerPage={10}
  location={
    Object {
      "hash": "",
      "pathname": "/0/spot",
      "search": "",
      "state": undefined,
    }
  }
  match={
    Object {
      "isExact": true,
      "params": Object {},
      "path": "/0/spot",
      "url": "localhost:8080/0/spot",
    }
  }
  newRowWrapper={[Function]}
  numOfFilteredRows={2}
  onSorterChange={[Function]}
  page={1}
  rowWrapper={[Function]}
  rowsInPage={
    Immutable.List [
      Object {
        "id": 1,
        "name": "Spot 1",
        "requiredSkillSet": Array [],
        "tenantId": 0,
        "version": 0,
      },
      Object {
        "id": 2,
        "name": "Spot 2",
        "requiredSkillSet": Array [
          Object {
            "id": 0,
            "name": "Skill 1",
            "tenantId": 0,
            "version": 0,
          },
        ],
        "tenantId": 0,
        "version": 0,
      },
    ]
  }
  sortByIndex={0}
  title="Trans(i18nKey=spots)"
/>
`;

exports[`Spots page should render correctly with no spots 1`] = `
<TheTable
  columns={
    Array [
      Object {
        "name": "Trans(i18nKey=name)",
        "sorter": [Function],
      },
      Object {
        "name": "Trans(i18nKey=requiredSkillSet)",
      },
    ]
  }
  filterText=""
  filteredRows={Immutable.List []}
  history={
    Object {
      "action": "PUSH",
      "block": [MockFunction],
      "createHref": [MockFunction],
      "go": [MockFunction],
      "goBack": [MockFunction],
      "goForward": [MockFunction],
      "length": 1,
      "listen": [MockFunction],
      "location": Object {
        "hash": "",
        "pathname": "/0/spot",
        "search": "",
        "state": undefined,
      },
      "push": [MockFunction],
      "replace": [MockFunction],
    }
  }
  isReversed={true}
  itemsPerPage={10}
  location={
    Object {
      "hash": "",
      "pathname": "/0/spot",
      "search": "",
      "state": undefined,
    }
  }
  match={
    Object {
      "isExact": true,
      "params": Object {},
      "path": "/0/spot",
      "url": "localhost:8080/0/spot",
    }
  }
  newRowWrapper={[Function]}
  numOfFilteredRows={0}
  onSorterChange={[Function]}
  page={1}
  rowWrapper={[Function]}
  rowsInPage={Immutable.List []}
  sortByIndex={0}
  title="Trans(i18nKey=spots)"
/>
`;

exports[`Spots page should render the editor correctly 1`] = `
<TableRow>
  <TableCell>
    <TextInput
      onChange={[Function]}
      value="Spot"
    />
  </TableCell>
  <TableCell>
    <MultiTypeaheadSelectInput
      emptyText="Trans(i18nKey=selectRequiredSkills)"
      onChange={[Function]}
      optionToStringMap={[Function]}
      options={
        Array [
          Object {
            "id": 0,
            "name": "Skill 1",
            "tenantId": 0,
            "version": 0,
          },
        ]
      }
      value={Array []}
    />
  </TableCell>
  <RowEditButtons
    isValid={true}
    onClose={[Function]}
    onSave={[Function]}
  />
</TableRow>
`;

exports[`Spots page should render the viewer correctly 1`] = `
<TableRow>
  <TableCell>
    <Flex>
      <FlexItem>
        <Text>
          Spot 2
        </Text>
      </FlexItem>
      <FlexItem>
        <Button
          onClick={[Function]}
          variant="link"
        >
          <ArrowIcon
            color="currentColor"
            noVerticalAlign={false}
            size="sm"
          />
        </Button>
      </FlexItem>
    </Flex>
  </TableCell>
  <TableCell>
    <ChipGroup
      aria-label="Chip group category"
      categoryName=""
      closeBtnAriaLabel="Close chip group"
      collapsedText="\${remaining} more"
      defaultIsOpen={false}
      expandedText="Show Less"
      isClosable={false}
      numChips={3}
      onClick={[Function]}
      onOverflowChipClick={[Function]}
      tooltipPosition="top"
    >
      <Chip
        className=""
        closeBtnAriaLabel="close"
        component="div"
        isOverflowChip={false}
        isReadOnly={true}
        key="Skill 1"
        onClick={[Function]}
        tooltipPosition="top"
      >
        Skill 1
      </Chip>
    </ChipGroup>
  </TableCell>
  <RowViewButtons
    onDelete={[Function]}
    onEdit={[Function]}
  />
</TableRow>
`;
